### ISSUE TITLE
Create a demo server to try out FreedomBox 

### ISSUE THREAD

**Issue by JosephKiranBabu** [freedombox/Plinth#862]
A demo version of FreedomBox that new users can try out will help a lot in adoption.
We can have a VM that gets redeployed every 1 hour.
Create a link in the freedombox.org home page pointing to the demo instance. Something like "try.freedombox.org" for new users to check it out.

**Comment by JoKeyser** [comment]
Hi @JosephKiranBabu , great idea. Such an infrastructure could probably also be used for some usability tests with real, naive users.

**Comment by mray** [comment]
Aside from possible security concerns this sounds like a good idea. Don't we essentially grant root rights to any plinth visitor? My gut says this may end badly - but I totally like the idea of a demo server. I remember the Sandstorm demo server being a very convincing experience...

**Comment by JoKeyser** [comment]
I agree with @mray, maybe some things have to be restricted in the VM. Probably SSH login should be off, to prevent installation of a mailserver for spam, or worse things. Maybe some proof that the user is human would be a good idea before logging in.

The point is to test interaction with the the Plinth interface, e.g. let the user install a service they care about, and they should be able to actually login with a client to see their own creation in action. They could even bring their own (sub-)domain, if they plan ahead and point it to that test-VM (DNS changes often take hours), so they could check out the (upcoming) Let's Encrypt functions, Monkeysphere, etc.

Which version should be run? Stable? Testing? Unstable? Git? Maybe the user can choose?

**Comment by nicorikken** [comment]
(discussed a bit with @JoKeyser ) I see 2 options of doing this:

1. Running a full FreedomBox instance per demo user (Docker container?)
    - Pro: admin access (installing apps), connecting a client
    - Con: more elaborate hosting (cost), security risk (spamming)
2. Running a mocked/simulated FreedomBox UI (similar behavior, but no apps are installed/uninstalled)
    - Pro: responsive (no real execution), secure, cheap hosting (single server)
    - Con: keeping simulation/mock up to date with changes (extra effort), can't connect clients

As an extension of 2 (even 1 perhaps), could be to record the actions of demo users if they agree to it upfront. That could provide us with ongoing usability information of real (new) people. This could be a full log of all the clicks in the UI. Also we could include a 'feedback' sidebar, or like a right-click menu so people can write a note at any moment.

**Comment by JoKeyser** [comment]
Running a full instance has the additional benefit of exposing bugs. Maybe the security issues can be mitigated by aggressive output filtering, while still allowing the few expected apps to send things (but I think testing the software with a client is a minority feature anyhow). Maybe the responsiveness can be improved by caching a local, limited repository.

I think a combination of a demo server with feedback capabilities would be really helpful. Thanks to Django, adding feedback can be achieved by middleware, see e.g. https://pypi.python.org/pypi/django-tellme

**Comment by JosephKiranBabu** [comment]
I've seen a few deployment models for demo servers.

1. Nextcloud seems to deploy a container for each demo user.
2. Sandcats has some kind of sandboxing mechanism where a user's changes are kept isolated. In case of Sandcats, they just pretend to install the application, while in case of FreedomBox that might not be satisfactory. 

Ideally, we should deploy one demo instance per user. Maybe one vagrant box per user? We can automate destruction of vagrant boxes that are older than a specified time limit.

**Comment by JoKeyser** [comment]
Another feature would be to provide translators feedback of their translated strings in-context, fully interactive. Probably for this purpose, an instance with everything installed would be best.

**Comment by JosephKiranBabu** [comment]
@JoKeyser Wouldn't that more be like a staging server? I think the demo server should run a somewhat stable version so that users don't face the regressions in our development version. Also, I think the translated strings are only committed once a day by Weblate. The turnaround time even for the staging server is high for translators. 🤐

**Comment by JoKeyser** [comment]
@JosephKiranBabu yep I think you're right, the demo server should be somewhat stable. So if Weblate commits rather infrequently, only they could show things in context, but that's probably way beyond their scope. However in Weblate, rather recent screenshots would be nice, maybe there's a way to automate that, e.g. with your Selenium magic? Anyways, sorry for the tangent, I guess we can live without that; it's not a real bottleneck, rather something nice-to-have.
